micro-rdk/src/common/webrtc/api.rs

```diff
@@ -349,88 +395,145 @@ impl WebRtcSignalingChannel {
-                            }
-                            answer_request::Stage::Heartbeat(_) => {
-                                log::debug!("received a heartbeat from the signaling server");
-                                continue;
                             }
-                            _ => {
-                                return Err(WebRtcError::SignalingError(
-                                    "unexpected stage".to_owned(),
-                                ))
+                        }
+                        return Ok(None);
+                    }
+                }
+            },
+            Either::Right(ref mut local_signaling) => {
+                match local_signaling.rx.recv().await {
+                    Err(RecvError) => Err(WebRtcError::SignalingDisconnected()),
+                    Ok(req) => {
+                        if let Some(update) = req.update {
+                            match update {
+                                call_update_request::Update::Candidate(c) => c
+                                    .candidate
+                                    .try_into()
+                                    .map_err(|_| WebRtcError::CannotParseCandidate)
+                                    .map(Option::Some),
+                                call_update_request::Update::Done(_) => {
+                                    // TODO(acm): Verify that the actual boolean in the `Done` doesn't make any difference
+                                    Ok(None)
+                                }
+                                call_update_request::Update::Error(e) => {
+                                    Err(WebRtcError::SignalingError(e.message))
+                                }
                             }
+                        } else {
+                            Ok(None)
                         }
                     }
-                    return Ok(None);
                 }
             }
         }
     }
+
     pub async fn send_done(&mut self, uuid: String) -> Result<(), WebRtcError> {
-        let answer = AnswerResponse {
-            uuid,
-            stage: Some(answer_response::Stage::Done(AnswerResponseDoneStage {})),
-        };
-        match self.signaling_tx.send_message(answer).await {
-            Err(_) => Err(WebRtcError::SignalingDisconnected()),
-            Ok(_) => Ok(()),
+        match self.signaling.as_mut() {
+            Either::Left(ref mut app_signaling) => {
+                let answer = AnswerResponse {
+                    uuid,
+                    stage: Some(answer_response::Stage::Done(AnswerResponseDoneStage {})),
+                };
+                match app_signaling.tx.send_message(answer).await {
+                    Err(_) => Err(WebRtcError::SignalingDisconnected()),
+                    Ok(_) => Ok(()),
+                }
+            }
+            Either::Right(ref mut local_signaling) => match local_signaling.tx.close() {
```

Comment:
This is actually buggy. We should not be closing the channel here, because we may race with a CallUpdate that comes after we have sent all of the information from our side.

Reply:
why aren't we using a similar pattern to the `AnswerStage` structure? If we did, could we then differentiate between "done" and "should close"?

Reply:
The `Answer*` types have additional states that we don't have access to as a pure `Call` machine. There actually isn't anything we need to do here, except fix it to not close the channel. The channel close should happen for `call_update_request::Update::Done(_)`, instead.